storage/aptosdb/src/pruner/state_kv_pruner_worker.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::pruner::{db_pruner::DBPruner, state_kv_pruner::StateKvPruner};
+use aptos_config::config::StateKvPrunerConfig;
+use aptos_logger::{
+    error,
+    prelude::{sample, SampleRate},
+};
+use aptos_types::transaction::Version;
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread::sleep,
+    time::Duration,
+};
+
+/// Maintains the state kv pruner and periodically calls the db_pruner's prune method to prune the DB.
+/// This also exposes API to report the progress to the parent thread.
+pub struct StateKvPrunerWorker {
+    /// The worker will sleep for this period of time after pruning each batch.
+    pruning_time_interval_in_ms: u64,
+    /// State kv pruner.
+    pruner: Arc<StateKvPruner>,
+    /// Max number of versions to prune per batch.
+    max_versions_to_prune_per_batch: u64,
+    /// Indicates whether the pruning loop should be running. Will only be set to true on pruner
+    /// destruction.
+    quit_worker: AtomicBool,
+}
+
+impl StateKvPrunerWorker {
+    pub(crate) fn new(
+        state_kv_pruner: Arc<StateKvPruner>,
+        state_kv_pruner_config: StateKvPrunerConfig,
+    ) -> Self {
+        Self {
+            pruning_time_interval_in_ms: if cfg!(test) { 100 } else { 1 },
+            pruner: state_kv_pruner,
+            max_versions_to_prune_per_batch: state_kv_pruner_config.batch_size as u64,
+            quit_worker: AtomicBool::new(false),
+        }
+    }
+
+    // Loop that does the real pruning job.
+    pub(crate) fn work(&self) {
+        while !self.quit_worker.load(Ordering::Relaxed) {
```

Comment:
are there multiple threads running the same StateKvPrunerWorker?

Reply:
No, there is only one worker. We use atomic variable here just because the stop_pruning signal is from other thread (not the worker thread).